src/arkode/arkode_forcingstep.c

```diff
@@ -0,0 +1,415 @@
+                          (tcur, ycur) or the start of the subsequent step i.e.,
+                          at (tn, yn) = (tcur, ycur) from the end of the last
+                          step
+
+     ARK_FULLRHS_OTHER -> called elsewhere (e.g. for dense output)
+
+  In ForcingStep, we accumulate the RHS functions in ARK_FULLRHS_OTHER mode.
+  Generally, inner steppers will not have the correct yn when this function is
+  called and will not be able to reuse a function evaluation since their state
+  resets at the next SUNStepper_Evolve call.
+  ----------------------------------------------------------------------------*/
+static int forcingStep_FullRHS(const ARKodeMem ark_mem, const sunrealtype t,
+                               const N_Vector y, const N_Vector f,
+                               SUNDIALS_MAYBE_UNUSED const int mode)
+{
+  ARKodeForcingStepMem step_mem = NULL;
+  const int retval = forcingStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return retval; }
+
+  SUNErrCode err = SUNStepper_FullRhs(step_mem->stepper[0], t, y,
+                                      ark_mem->tempv1, SUN_FULLRHS_OTHER);
+  if (err != SUN_SUCCESS)
+  {
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_RHSFUNC_FAILED, t);
+    return ARK_RHSFUNC_FAIL;
+  }
+
+  /* TODO(SBR): this *may* be able to use SUN_FULLRHS_END mode in some cases
+   * but we need to be certain the state of the SUNStepper is consistent with
+   * the outer forcing method */
+  err = SUNStepper_FullRhs(step_mem->stepper[1], t, y, f, SUN_FULLRHS_OTHER);
+  if (err != SUN_SUCCESS)
+  {
+    arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, __LINE__, __func__, __FILE__,
+                    MSG_ARK_RHSFUNC_FAILED, t);
+    return ARK_RHSFUNC_FAIL;
+  }
+  N_VLinearSum(SUN_RCONST(1.0), f, SUN_RCONST(1.0), ark_mem->tempv1, f);
+
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  This routine performs a single step of the forcing method.
+  ---------------------------------------------------------------*/
+static int forcingStep_TakeStep(const ARKodeMem ark_mem,
+                                sunrealtype* const dsmPtr, int* const nflagPtr)
+{
+  ARKodeForcingStepMem step_mem = NULL;
+  int retval = forcingStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return retval; }
+
+  *nflagPtr = ARK_SUCCESS; /* No algebraic solver */
+  *dsmPtr   = ZERO;        /* No error estimate */
+
+  const SUNStepper s0    = step_mem->stepper[0];
+  const sunrealtype tout = ark_mem->tn + ark_mem->h;
+  sunrealtype tret       = 0;
+
+  /* Evolve stepper 0 on its own */
+  SUNErrCode err = SUNStepper_Reset(s0, ark_mem->tn, ark_mem->yn);
+  if (err != SUN_SUCCESS) { return ARK_SUNSTEPPER_ERR; }
+  err = SUNStepper_SetStepDirection(s0, ark_mem->h);
+  if (err != SUN_SUCCESS) { return ARK_SUNSTEPPER_ERR; }
```

Comment:
Suppose forcingstep is used to integrate forward for some time then `ARKodeReset` and `ARKodeSetStepDirection(..., -1)` are called on the forcingstep memory. Before evolving again (now backward in time), the partition integrators need to reverse direction. We can either have forcingstep do it or require the user to do it. Since splittingstep internally must call `SUNStepper_SetStepDirection` to handle direction changes, it made sense for forcingstep to do the same.

Reply:
If this makes sense, I'll add what I wrote as a comment in the code.

Reply:
The user should have to call `ARKodeSetStepDirection` if they want to change the integration direction but they shouldn't have to call it on each of the `SUNSteppers` individually just the "overall" `arkode_mem`. For this steppers could provide a `setstepdirection` function to ARKODE that `ARKodeSetStepDirection` will call (if it exists) to perform any additional actions needed. So the ForcingStep version of this function would call `SUNStepper_SetStepDirection` on each `SUNStepper` (and similarly for SpilttingStep and MRIStep) while ERKStep, ARKStep, etc. would not need to set this function pointer in the ARKODE memory.

This also makes me realize we that we should add a check in `ARKodeSteSetDirection` to ensure it's only called after `Reset` (or before the first `Evolve` call). This can be done by checking if `ark_mem->initialized` is true like in `ARKodeSetInterpolantType`.